internal/blobstore/multipart.go

```diff
@@ -0,0 +1,226 @@
+package blobstore
+
+import (
+	"fmt"
+	"gopkg.in/errgo.v1"
+	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+	"io"
+	"time"
+)
+
+var (
+	maxParts          = 400
+	minPartSize int64 = 5 * 1024 * 1024
+)
+
+var ErrNotFound = errgo.New("blob not found")
+
+// uploadDoc describes the record that's held
+// for a pending multipart upload.
+type uploadDoc struct {
+	// Id holds the upload id. The blob for each
+	// part in the underlying blobstore will be named
+	// $id/$partnumber.
+	Id string `bson:"_id"`
+
+	// Hash holds the SHA384 hash of all the
+	// concatenated parts. It is empty until
+	// after FinishParts is called.
+	Hash string `bson:"hash,omitempty"`
+
+	// Expires holds the expiry time of the upload.
+	Expires time.Time
+
+	// Parts holds all the currently uploaded parts.
+	Parts []*uploadPart
+}
+
+// uploadPart represents one part of an on-going upload.
+type uploadPart struct {
+	// Hash holds the SHA384 hash of the part.
+	Hash string
+	// Size holds the size of the part.
+	Size int64
+	// Complete holds whether the part has been
+	// successfully uploaded.
+	Complete bool
+}
+
+// NewParts created a new multipart entry to track
+// a multipart upload. It returns an uploadId that
+// can be used to refer to it.
+func (s *Store) NewParts(expires time.Time) (uploadId string, err error) {
+	// TODO this makes an upload id that's 78 bytes.
+	// A base64-encoded 256 bit random number would
+	// only be 45 bytes and secure enough that we could
+	// probably dispense with ownership checking.
+	uploadId = fmt.Sprintf("%x", bson.NewObjectId())
+	if err := s.uploadC().Insert(uploadDoc{
+		Id:      uploadId,
+		Expires: expires,
+	}); err != nil {
+		return "", errgo.Notef(err, "cannot create new upload")
+	}
+	return uploadId, nil
+}
+
+// PutPart uploads a part to the given upload id. The part number
+// is specified with the part parameter; its content will be read from r
+// and is expected to have the given size and hex-encoded SHA384 hash.
+// A given part may not be uploaded more than once with a different hash.
+//
+// If the upload id was not found (for example, because it's expired),
+// PutPart returns an error with an ErrNotFound cause.
+func (s *Store) PutPart(uploadId string, part int, r io.Reader, size int64, hash string) error {
+	if part < 0 {
+		return errgo.Newf("negative part number")
+	}
+	if part >= maxParts {
+		return errgo.Newf("part number %d too big (maximum %d)", part, maxParts-1)
+	}
+	uploadc := s.uploadC()
+	var udoc uploadDoc
+	if err := uploadc.FindId(uploadId).One(&udoc); err != nil {
+		if err == mgo.ErrNotFound {
+			return errgo.WithCausef(nil, ErrNotFound, "upload id %q not found", uploadId)
+		}
+		return errgo.Notef(err, "cannot get upload id %q", uploadId)
+	}
+	if len(udoc.Parts) > 0 && part >= len(udoc.Parts) {
```

Comment:
Can't we do this check regardless of whether the new part is > the total number of uploaded parts?
What if you upload part 3 first, with not enough bytes, then part 0? At that point len(udoc.Parts) > 0 (1 > 0) but part < len(udoc.Parts) (0 < 1), so part 3 is never checked?